src/wallet/KintoWallet.sol

```diff
@@ -351,34 +373,26 @@ contract KintoWallet is Initializable, BaseAccount, TokenCallbackHandler, IKinto
         dest.functionCallWithValue(func, value);
     }
 
-    // Function to extract the first target contract
-    function _getAppContract(bytes calldata callData) private view returns (address) {
-        // Extract the function selector from the callData
-        bytes4 selector = bytes4(callData[:4]);
+    // extracts `target` contract from the callData
+    // @dev when calling executeBatch, we use the first element of the targets to determine the app
+    function _getTarget(bytes calldata callData) private view returns (address target) {
+        bytes4 selector = bytes4(callData[:4]); // extract the function selector from the callData
 
-        // Compare the selector with the known function selectors
         if (selector == IKintoWallet.executeBatch.selector) {
-            // Decode callData for executeBatch
+            // decode executeBatch callData
             (address[] memory targets,,) = abi.decode(callData[4:], (address[], uint256[], bytes[]));
-            address lastTargetContract = appRegistry.getSponsor(targets[targets.length - 1]);
+            require(targets.length > 0, "KW: no targets");
+
+            // find the first non-wallet contract in the batch to determine the target
```

Comment:
we were previously using the concept of the "last target contract" but i think that was wrong since you could have an executeBatch for example with this:

- wallet.whitelistApp
- counter.increment()
- counter.increment()
- wallet.whitelistApp

this snceario SHOULD work but would fail as it was since the last target contract wouldn't be the app but could be some call to the wallet

Reply:
this scenario should work but the problem is who pays for it, with the current implementation the user would pay for it so it would be better if the app always puts the app txs last, in this case the app is counter

Reply:
hmm.. let me check if i understand:

wallet.whitelistApp -> should be paid by **user**
counter.increment() -> should be paid by **app**
counter.increment() -> should be paid by **app**
wallet.whitelistApp -> should be paid by **user**

is this what you mean? that only the "app" calls should be paid by the app? if so, i would probably just ensure that all the actions are made to the same target and that the target is the app or any children (and also any sponsored contract as well)? and, if there's a call to wallet then the whole executeBatch should fail

Reply:
we talked about this offline. App is happy to pay for everything as long as there are not that many wallet ops and the other calls are either children or sponsored